subsys/shell/modules/kernel_service.c

```diff
@@ -333,8 +271,269 @@ static int cmd_kernel_stacks(const struct shell *sh,
+	if (err != 0) {
+		shell_error(sh, "Unable to parse thread ID %s (err %d)", argv[1], err);
+		return err;
+	}
+
+	cpu = (int)shell_strtol(argv[2], 10, &err);
+	if (err != 0) {
+		shell_error(sh, "Unable to parse CPU ID %s (err %d)", argv[2], err);
+		return err;
+	}
+
+	rc = k_thread_cpu_mask_disable(thread, cpu);
+	if (rc != 0) {
+		shell_error(sh, "Failed - %d", rc);
+	} else {
+		shell_print(sh, "%p %s cpu_mask: 0x%x", (void *)thread, thread->name,
+			    thread->base.cpu_mask);
+	}
+
+	return rc;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_kernel_thread_mask,
+	SHELL_CMD_ARG(clear, NULL,
+		      "Sets all CPU enable masks to zero.\n"
+		      "Usage: kernel thread mask clear <thread ID>",
+		      cmd_kernel_thread_mask_clear, 2, 0),
+	SHELL_CMD_ARG(enable_all, NULL,
+		      "Sets all CPU enable masks to one.\n"
+		      "Usage: kernel thread mask enable_all <thread ID>",
+		      cmd_kernel_thread_mask_enable_all, 2, 0),
+	SHELL_CMD_ARG(enable, NULL,
+		      "Enable thread to run on specified CPU.\n"
+		      "Usage: kernel thread mask enable <thread ID> <CPU ID>",
+		      cmd_kernel_thread_mask_enable, 3, 0),
+	SHELL_CMD_ARG(disable, NULL,
+		      "Prevent thread to run on specified CPU.\n"
+		      "Usage: kernel thread mask disable <thread ID> <CPU ID>",
+		      cmd_kernel_thread_mask_disable, 3, 0),
+	SHELL_SUBCMD_SET_END /* Array terminated. */
+);
+
+static int cmd_kernel_thread_pin(const struct shell *sh,
+				  size_t argc, char **argv)
+{
+	ARG_UNUSED(argc);
+
+	int cpu, err = 0;
+	struct k_thread *thread;
+
+	thread = UINT_TO_POINTER(shell_strtoull(argv[1], 16, &err));
+	if (err != 0) {
+		shell_error(sh, "Unable to parse thread ID %s (err %d)", argv[1], err);
+		return err;
+	}
+
+	if (!thread_valid(thread)) {
+		shell_error(sh, "Invalid thread id %p", (void *)thread);
+		return -EINVAL;
+	}
+
+	cpu = shell_strtoul(argv[2], 10, &err);
+	if (err != 0) {
+		shell_error(sh, "Unable to parse CPU ID %s (err %d)", argv[2], err);
+		return err;
+	}
+
+	shell_print(sh, "Pinning %p %s to CPU %d", (void *)thread, thread->name, cpu);
+	err = k_thread_cpu_pin(thread, cpu);
+	if (err != 0) {
+		shell_error(sh, "Failed - %d", err);
+	} else {
+		shell_print(sh, "%p %s cpu_mask: 0x%x", (void *)thread, thread->name,
+			    thread->base.cpu_mask);
+	}
+
+	return err;
+}
+#endif /* CONFIG_SCHED_CPU_MASK */
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_kernel_thread,
+#if defined(CONFIG_INIT_STACKS) && defined(CONFIG_THREAD_STACK_INFO) && \
+		defined(CONFIG_THREAD_MONITOR)
+	SHELL_CMD(stacks, NULL, "List threads stack usage.", cmd_kernel_thread_stacks),
+	SHELL_CMD(list, NULL, "List kernel threads.", cmd_kernel_thread_list),
+#endif
+#if defined(CONFIG_ARCH_HAS_STACKWALK)
+	SHELL_CMD_ARG(unwind, NULL, "Unwind a thread.", cmd_kernel_thread_unwind, 1, 1),
+#endif /* CONFIG_ARCH_HAS_STACKWALK */
+#ifdef CONFIG_SCHED_CPU_MASK
+	SHELL_CMD_ARG(mask, &sub_kernel_thread_mask, "Configure thread CPU mask affinity.",
```

Comment:
Did you consider using`SHELL_COND_CMD_ARG_REGISTER` instead of `ifdef`?

Reply:
No, ain't aware of that, I pushed another version with `SHELL_COND_CMD_ARG`, PTAL

Reply:
Reverted back from `SHELL_COND_CMD_ARG` to `SHELL_CMD_ARG` because the former requires the command to be there and thus the `#ifdef` can't be avoided, they just relocated